File: leetcode/go/122.best-time-to-buy-and-sell-stock-ii.go
func maxProfit(prices []int) int {
    maxp, minp, ans := prices[0], prices[0], 0
    for _, p := range prices {
        if p < maxp {
            ans += (maxp - minp)
            minp = p
            maxp = p
        } else if p > maxp {
            maxp = p
        }
        // fmt.Println(p, minp, maxp, ans)
    }
    if maxp > minp {
        ans += (maxp - minp)
    }
    return ans
}

/* */

func maxProfit(prices []int) int {
    ans := 0
    for i:=0; i<len(prices)-1; i++ {
        if prices[i] < prices[i+1] {
            ans += (prices[i+1]-prices[i])
        }
    }
    return ans
}

/* */

func maxProfit(prices []int) int {
    n := len(prices)
    dp := make([][]int, n+1)
    dp[n] = []int{0, 0}
    for i:=n-1; i > -1; i-- {
        dp[i] = []int{0, 0}
        for j := range 2 {
            if j == 1 {
                dp[i][j] = max(dp[i+1][1], dp[i][0]-prices[i])
            } else {
                dp[i][j] = max(prices[i] + dp[i+1][1], dp[i+1][0])
            }
        }
    }
    // for _, row := range dp {
    //     fmt.Println(row)
    // }
    return dp[0][1]
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: leetcode/rust-leetcode/src/122.best-time-to-buy-and-sell-stock-ii.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut total = 0i32;
        let mut minp = prices[0];
        let mut maxp = prices[0];
        for p in prices {
            if p < maxp {
                total += (maxp - minp);
                minp = p;
                maxp = p;
            } else if p > maxp {
                maxp = p;
            }
        }
        if maxp > minp {
            total += (maxp - minp);
        }
        total
    }
}

/*
*/

use std::cmp;

impl Solution2 {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let n: usize = prices.len();
        let mut dp: Vec<Vec<i32>> = vec![vec![0; 2]; n+1];
        dp[n][0] = 0;
        dp[n][1] = 0;
        for i in (0..n).rev() {
            for j in 0..2 {
                if j == 1 {
                    dp[i][j] = cmp::max(-prices[i] + dp[i+1][0], dp[i+1][1]);
                } else {
                    dp[i][j] = cmp::max(prices[i] + dp[i+1][1], dp[i+1][0]);
                }
            }
        }
        dp[0][1]
    }
}